crates/sui-package-resolver/src/lib.rs

```diff
@@ -448,6 +477,122 @@ impl<S: PackageStore> Resolver<S> {
             .map(|t| t.as_ref().and_then(|t| layouts.get(t).cloned()))
             .collect())
     }
+
+    /// Resolves a runtime address in a `ModuleId` to a storage `ModuleId` according to the linkage
+    /// table in the `context` which must refer to a package.
+    /// * Will fail if the wrong context is provided, i.e., is not a package, or
+    ///   does not exist.
+    /// * Will fail if an invalid `context` is provided for the `location`, i.e., the package at
+    ///   `context` does not contain the module that `location` refers to.
+    pub async fn resolve_module_id(
+        &self,
+        module_id: ModuleId,
+        context: AccountAddress,
+    ) -> Result<ModuleId> {
+        let package = self.package_store.fetch(context).await?;
+        let storage_id = package.relocate(*module_id.address())?;
+        Ok(ModuleId::new(storage_id, module_id.name().to_owned()))
+    }
+
+    /// Resolves an abort code following the clever error format to a `CleverError` enum.
+    /// The `module_id` must be the storage ID of the module (which can e.g., be gotten from the
+    /// `resolve_module_id` function) and not the runtime ID.
+    ///
+    /// If the `abort_code` is not a clever error (i.e., does not follow the tagging and layout as
+    /// defined in `ErrorBitset`), this function will return `None`.
+    ///
+    /// In the case where it is a clever error but only a line number is present (i.e., the error
+    /// is the result of an `assert!(<cond>)` source expression) a `CleverError::LineNumberOnly` is
+    /// returned. Otherwise a `CleverError::CompleteError` is returned.
+    ///
+    /// If for any reason we are unable to resolve the abort code to a `CleverError`, this function
+    /// will return `None`.
+    pub async fn resolve_clever_error(
+        &self,
+        module_id: ModuleId,
+        abort_code: u64,
+    ) -> Option<CleverError> {
+        let Some(bitset) = ErrorBitset::from_u64(abort_code) else {
+            return None;
+        };
+
+        let package = self.package_store.fetch(*module_id.address()).await.ok()?;
+        let module = package.module(module_id.name().as_str()).ok()?.bytecode();
+
+        let source_line_number = bitset.line_number()?;
+
+        // We only have a line number in our clever error, so return early.
+        if bitset.identifier_index().is_none() && bitset.constant_index().is_none() {
+            return Some(CleverError::LineNumberOnly {
+                module_id,
+                source_line_number,
+            });
+        }
```

Comment:
What's the right thing to do if only one of these `is_none`? Does that imply that it's not really a clever error? I think today, we'll still try and read a constant from the pool in that case, which we probably don't want -- better to be explicit here:

```suggestion
        if bitset.identifier_index().is_none() && bitset.constant_index().is_none() {
            return Some(CleverError::LineNumberOnly {
                module_id,
                source_line_number,
            });
        } else if bitset.identifier_index().is_none() || bitset.constant_index().is_none() {
            return None;
        }
```

Reply:
Yea, we should never have a clever error (at least under the current design) where you would have one without the other, so adding this extra check makes sense to me